### PR TITLE
Add `beforeError` hook

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes https://github.com/sindresorhus/gh-got/blob/master/index.js

Instead using handler we could do:

```js
hooks: {
	beforeRequest: [
		options => {
			if (options.token) {
				options.headers.authorization = options.headers.authorization || `token ${options.token}`;
			}
		}
	],
	afterResponse: [
		response => {
			response.rateLimit = getRateLimit(response);
			return response;
		}
	],
	onError: [
		error => {
			const {response} = error;

			if (response && response.body) {
				error.name = 'GitHubError';
				error.message = `${response.body.message} (${error.statusCode})`;
			}

			if (response) {
				error.rateLimit = getRateLimit(response);
			}

			return error;
		}
	]
}
```

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [x] If it's a new feature, I have included documentation updates.

### CODE REVIEW

**Comment by sindresorhus** [comment]
> Fixes sindresorhus/gh-got:index.js@master

I wouldn't say it fixes that issue. This just works around it. We still need to properly fix `handler`.

File: readme.md [hunk 1]
```diff
@@ -447,6 +447,38 @@ const instance = got.extend({
...
+				const {response} = error;
+
+				if (response && response.body) {
+					error.name = 'GitHubError';
+					error.message = `${response.body.message} (${error.statusCode})`;
+				}
+
+				if (response) {
+					error.rateLimit = getRateLimit(response);
+				}
```
Comment (sindresorhus) [inline readme.md]: I would drop this part. The example is already complete without this.

**Comment by sindresorhus** [comment]
Does the hook catch absolutely *all* errors? If not, it should be documented what it doesn't catch. For example, I don't think it currently catches input validation errors (Neither should it IMHO).

**Comment by sindresorhus** [comment]
Wouldn't `beforeError` be a more correct name for this?

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by szmarczak** [comment]
done

**Pull request merged**